### PR TITLE
Neighborhood: add all client signals

### PR DESCRIPTION
**Pull request by molly-moen**
Add all the signals neighborhood needs to send to the client for various client-facing actions. Move had been implemented previously. This PR adds:
- INITIALIZE
- PAINT
- REMOVE_PAINT
- TAKE_PAINT
- HIDE_PAINTER
- SHOW_PAINTER

These signals will be used on javalab to invoke various painter actions.
As a part of this PR, also updated the variable `visible` on `UserFileData` to be `isVisible` since we have updated the variable on the javalab side.

### CODE REVIEW

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java [hunk 1]
```diff
@@ -1,6 +1,18 @@
...
+  // Paint the current location
+  PAINT,
+  // Remove all paint from current location
+  REMOVE_PAINT,
+  // Take paint from the bucket
+  TAKE_PAINT,
+  // Hide the painter on the screen
+  HIDE_PAINTER,
+  // Show the painter on the screen
+  SHOW_PAINTER
```
Comment (jmkulwik) [inline org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:17]: I think we also need TURN_LEFT

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java [hunk 2]
```diff
@@ -1,6 +1,18 @@
 package org.code.neighborhood;
 
 public enum NeighborhoodSignalKey {
+  // Initialize a new painter
+  INITIALIZE,
```
Comment (jmkulwik) [inline org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:5]: Is this initializing the whole maze or an individual painter? If it's the latter, a name such as INITIALIZE_PAINTER may make more sense.

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java (hunk 1 above)
Comment (jmkulwik) [inline org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:17]: One item that wasn't on the list but we'll also need to add is HIDE_BUCKETS to make all the buckets on a level disappear.

Comment (molly-moen) [reply org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:17]: good catch

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java (hunk 2 above)
Comment (molly-moen) [reply org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:5]: it's a single painter, I will update

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java (hunk 1 above)
Comment (jmkulwik) [reply org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:17]: Definitely can be a follow-up. Looks like this was missed even on the API description.

Comment (molly-moen) [reply org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/NeighborhoodSignalKey.java:17]: it's a quick update, I'll add it

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/Grid.java [hunk 3]
```diff
@@ -41,12 +42,16 @@ public GridSquare getSquare(int x, int y) {
...
-    System.out.println("You hid the buckets");
+    NeighborhoodOutputHandler.sendMessage(
+            new NeighborhoodSignalMessage(NeighborhoodSignalKey.HIDE_BUCKETS, new HashMap<>())
+    );
   }
 
   // Displays all buckets on the screen
   public void showBuckets() {
-    System.out.println("You displayed the buckets");
+    NeighborhoodOutputHandler.sendMessage(
```
Comment (jmkulwik) [inline org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/Grid.java:52]: [Non-blocking comment] I think Hannah wanted students to be able to call this directly from the Painter object. We'll want to double check the use case. Definitely not blocking for this PR as that could take a couple cycles to figure out.

**Review by jmkulwik — APPROVED** [review]
Nice work!

**Review by jamescodeorg — COMMENTED** [review]
Where can I find Maze's "interface" on the client side?

I'm wondering if it would be better to send the resolved state rather than the action (e.g. SET_DIRECTION instead of TURN_LEFT or TURN_RIGHT) but I guess it depends a bit on what the interface is on the client side?

Also, Hannah mentioned that students will be asked to extend Painter with methods such as turnRight().  Would that work by having painter turn left 3 times?  Or will there be a way for students to extend the set of commands/messages?

**Comment by molly-moen** [comment]
@jamescodeorg the controller we will use to access maze functions is [here](https://github.com/code-dot-org/maze/blob/main/src/mazeController.js), let me know if that's enough info!
For the turn question, we could go either way. I kept it aligned with what the command was on the server-side, but from the client side we will be calling turn with a direction (see function [here](https://github.com/code-dot-org/maze/blob/main/src/mazeController.js#L247)), so SET_DIRECTION would also make sense. However, students will not be extending the set of commands/messages--this logic is all hidden from them. So their turnRight() would be turnLeft() 3 times.

File: org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/Painter.java [hunk 4]
```diff
@@ -181,4 +190,22 @@ private boolean isValidMovement(Direction movementDirection) {
...
+    return details;
+  }
+
+  private void sendOutputMessage(NeighborhoodSignalKey signalKey, HashMap<String, String> details) {
+    NeighborhoodOutputHandler.sendMessage(new NeighborhoodSignalMessage(signalKey, details));
+  }
+
+  private void sendInitializationMessage() {
+    HashMap<String, String> initDetails = this.getSignalDetails();
+    initDetails.put("direction", this.direction.getDirectionString());
```
[the remark this answers is not in the document]
Comment (molly-moen) [reply org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/Painter.java:206]: painter id is added in the `getSignalDetails` method

**Comment by jamescodeorg** [comment]
Got it, thanks!  Lots of interesting questions around how to keep the back-end and front-end in sync, it will be interesting to see how this evolves!

**Pull request merged**